File: eval-reviews/2026-09-12-chained/query-null/python/fable/r1/starter/binder.py
"""Resolve all names and types before optimization or row evaluation."""
from model import Plan, require
from parser import AGGREGATES


def walk(e):
    yield e
    for a in e.args: yield from walk(a)


def bind(q, database):
    aliases = [a for _, a in q.select]
    require(len(set(aliases)) == len(aliases), 'DUPLICATE_ALIAS')
    require(len(set(a for _, a in q.sources)) == len(q.sources), 'DUPLICATE_ALIAS')
    tables, columns = [], []
    for si, (name, alias) in enumerate(q.sources):
        require(name in database, 'UNKNOWN_TABLE')
        t = database[name]
        tables.append(t)
        # The null-supplied side of a LEFT JOIN may be padded with NULL regardless of declaration.
        padded = si > 0 and q.kinds[si - 1] == 'LEFT'
        columns.extend((alias, c['name'], c['type'], si, c['nullable'] or padded) for c in t['columns'])
        if si: check(q.joins[si - 1], columns, False, False, True)
    for e in q.groups: check(e, columns, False)
    keys = [e.index for e in q.groups]
    require(len(set(keys)) == len(keys), 'INVALID_AGGREGATION')
    if q.where: check(q.where, columns, False, False, True)
    for e, _ in q.select: check(e, columns, True)
    if q.having: check(q.having, columns, True, False, True)
    roots = [e for e, _ in q.select] + ([q.having] if q.having else [])
    aggregate = bool(q.groups) or any(e.op in AGGREGATES for root in roots for e in walk(root))
    require(not q.having or aggregate, 'INVALID_AGGREGATION')
    if aggregate:
        for root in roots: grouped(root, keys)
    q.order = [(aliases.index(a) if a in aliases else -1, d, nf) for a, d, nf in q.order]
    require(all(i >= 0 for i, _, _ in q.order), 'UNKNOWN_COLUMN')
    return Plan(q, tables, [[] for _ in tables], aggregate)


def grouped(e, keys):
    if e.op in AGGREGATES: return
    require(e.op != 'col' or e.index in keys, 'INVALID_AGGREGATION')
    for a in e.args: grouped(a, keys)


def check(e, columns, allow, inside=False, predicate=False):
    """Resolve names and compute type plus a conservative nullable flag.

    'null' is the type of an untyped NULL literal; it unifies with any type
    required by its context."""
    if e.op == 'col':
        qualifier, name = e.value
        matches = [(i, c) for i, c in enumerate(columns) if c[1] == name and (not qualifier or c[0] == qualifier)]
        require(bool(matches), 'UNKNOWN_COLUMN')
        require(len(matches) == 1, 'AMBIGUOUS_COLUMN')
        e.index, c = matches[0]
        e.type, e.source, e.nullable = c[2], c[3], c[4]
    elif e.op == 'lit':
        e.nullable = e.value is None
    else:
        agg = e.op in AGGREGATES
        require(not agg or (allow and not inside), 'INVALID_AGGREGATION')
        for a in e.args: check(a, columns, allow, inside or agg)
        ts = [a.type for a in e.args]
        typed = [t for t in ts if t != 'null']
        e.nullable = any(a.nullable for a in e.args)
        if e.op == 'COUNT':
            e.type, e.nullable = 'int', False
        elif e.op in ('SUM', '+', '-', '*'):
            require(all(t == 'int' for t in typed), 'TYPE_ERROR')
            e.type = 'int'
            if e.op == 'SUM': e.nullable = True
        elif e.op in ('MIN', 'MAX'):
            require(ts[0] in ('int', 'text', 'null'), 'TYPE_ERROR')
            e.type, e.nullable = ts[0], True
        elif e.op in ('NOT', 'AND', 'OR'):
            require(all(t == 'bool' for t in typed), 'TYPE_ERROR')
            e.type = 'bool'
        elif e.op in ('ISNULL', 'ISNOTNULL'):
            e.type, e.nullable = 'bool', False
        elif e.op == 'COALESCE':
            require(all(t == typed[0] for t in typed), 'TYPE_ERROR')
            e.type = typed[0] if typed else 'null'
            e.nullable = all(a.nullable for a in e.args)
        else:
            require(all(t == typed[0] for t in typed) and (not typed or typed[0] != 'bool' or e.op in ('=', '<>')), 'TYPE_ERROR')
            e.type = 'bool'
    require(not predicate or e.type in ('bool', 'null'), 'TYPE_ERROR')

File: eval-reviews/2026-09-12-chained/query-null/typescript/fable/r1/starter/binder.ts
/** Binding and static validation precede all rewrites, even on empty input. */
import { aggregates, requireThat as need } from "./model.ts";
import type { Expr, Query, Plan, Table, ScalarType } from "./model.ts";
interface BoundColumn {
  qualifier: string;
  name: string;
  type: ScalarType;
  nullable: boolean;
  source: number;
}
export function* walk(e: Expr): Generator<Expr> {
  yield e;
  for (const a of e.args) yield* walk(a);
}
const comparisons = new Set(["=", "<>", "<", "<=", ">", ">="]);
/**
 * Type-check and annotate `e`. `padded` holds the sources whose columns may
 * have been NULL-padded by a LEFT JOIN at this point of the query; inside the
 * ON clause of a join the right-hand table is not yet padded.
 */
function check(
  e: Expr,
  columns: BoundColumn[],
  allow: boolean,
  padded: Set<number>,
  inside = false,
  predicate = false,
): void {
  if (e.op === "col") {
    const [qualifier, name] = e.value as [string, string];
    const matches = columns
      .map((c, i) => ({ c, i }))
      .filter(
        ({ c }) => c.name === name && (!qualifier || c.qualifier === qualifier),
      );
    need(matches.length > 0, "UNKNOWN_COLUMN");
    need(matches.length === 1, "AMBIGUOUS_COLUMN");
    const { c, i } = matches[0];
    e.index = i;
    e.type = c.type;
    e.source = c.source;
    e.nullable = c.nullable || padded.has(c.source);
  } else if (e.op === "lit") {
    e.nullable = e.value === null;
  } else {
    const agg = aggregates.has(e.op);
    need(!agg || (allow && !inside), "INVALID_AGGREGATION");
    for (const a of e.args) check(a, columns, allow, padded, inside || agg);
    // Untyped NULL literals (type undefined) adopt whatever type is required.
    const ts = e.args
      .map((a) => a.type)
      .filter((t): t is ScalarType => t !== undefined);
    const uniform = ts.every((t) => t === ts[0]);
    const anyNullable = e.args.some((a) => a.nullable);
    if (e.op === "COUNT") {
      e.type = "int";
      e.nullable = false;
    } else if (["SUM", "+", "-", "*"].includes(e.op)) {
      need(
        ts.every((t) => t === "int"),
        "TYPE_ERROR",
      );
      e.type = "int";
      e.nullable = e.op === "SUM" || anyNullable;
    } else if (["MIN", "MAX"].includes(e.op)) {
      need(
        ts.every((t) => t !== "bool"),
        "TYPE_ERROR",
      );
      e.type = ts[0];
      e.nullable = true;
    } else if (["NOT", "AND", "OR"].includes(e.op)) {
      need(
        ts.every((t) => t === "bool"),
        "TYPE_ERROR",
      );
      e.type = "bool";
      e.nullable = anyNullable;
    } else if (e.op === "IS NULL" || e.op === "IS NOT NULL") {
      e.type = "bool";
      e.nullable = false;
    } else if (e.op === "COALESCE") {
      need(uniform, "TYPE_ERROR");
      e.type = ts[0];
      e.nullable = e.args.every((a) => a.nullable);
    } else if (comparisons.has(e.op)) {
      need(
        uniform && (ts[0] !== "bool" || ["=", "<>"].includes(e.op)),
        "TYPE_ERROR",
      );
      e.type = "bool";
      e.nullable = anyNullable;
    } else throw new Error(`Unknown operator ${e.op}`);
  }
  need(!predicate || e.type === undefined || e.type === "bool", "TYPE_ERROR");
}
function grouped(e: Expr, keys: Set<number>): void {
  if (aggregates.has(e.op)) return;
  need(e.op !== "col" || keys.has(e.index!), "INVALID_AGGREGATION");
  for (const a of e.args) grouped(a, keys);
}
export function bind(q: Query, database: Map<string, Table>): Plan {
  const aliases = q.select.map(([, a]) => a);
  need(new Set(aliases).size === aliases.length, "DUPLICATE_ALIAS");
  need(
    new Set(q.sources.map(([, a]) => a)).size === q.sources.length,
    "DUPLICATE_ALIAS",
  );
  const tables: Table[] = [],
    columns: BoundColumn[] = [];
  const padded = new Set<number>();
  q.sources.forEach(([name, qualifier], source) => {
    const t = database.get(name);
    need(t, "UNKNOWN_TABLE");
    tables.push(t);
    columns.push(...t.columns.map((c) => ({ ...c, qualifier, source })));
    if (source) {
      check(q.joins[source - 1], columns, false, padded, false, true);
      if (q.outer[source - 1]) padded.add(source);
    }
  });
  q.groups.forEach((e) => check(e, columns, false, padded));
  const keys = new Set(q.groups.map((e) => e.index!));
  need(keys.size === q.groups.length, "INVALID_AGGREGATION");
  if (q.where) check(q.where, columns, false, padded, false, true);
  q.select.forEach(([e]) => check(e, columns, true, padded));
  if (q.having) check(q.having, columns, true, padded, false, true);
  const roots = [...q.select.map(([e]) => e), ...(q.having ? [q.having] : [])];
  const nodes = roots.flatMap((e) => [...walk(e)]);
  const aggregate =
    q.groups.length > 0 || nodes.some((e) => aggregates.has(e.op));
  need(!q.having || aggregate, "INVALID_AGGREGATION");
  if (aggregate) roots.forEach((e) => grouped(e, keys));
  q.order = q.order.map(([a, desc, nullsFirst]) => {
    const i = aliases.indexOf(a as string);
    need(i >= 0, "UNKNOWN_COLUMN");
    return [i, desc, nullsFirst];
  });
  return { query: q, tables, filters: tables.map(() => []), aggregate };
}
